presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/CteUtils.java

```diff
@@ -57,14 +73,42 @@ public static Integer getCtePartitionIndex(List<VariableReferenceExpression> out
         throw new PrestoException(GENERIC_INTERNAL_ERROR, "No Partitioning index found");
     }
 
-    /*
-        Currently, Hive bucketing does not support the Presto type 'ROW'.
-    */
     public static boolean isVariableMaterializable(VariableReferenceExpression var)
     {
-        return !(var.getType() instanceof RowType);
+        return isPrestoTypeBucketable(var.getType());
     }
 
+    /*
+       Currently, Hive bucketing does not support bucketing on these Presto types and user defined types
+       CHAR, DECIMAL, VARBINARY, ROW
+   */
+    public static boolean isPrestoTypeBucketable(Type type)
```

Comment:
Can we think of a way to do this that doesn't require engine changes that presume the underlying connector's capabilities?  I'm just thinking how this might work if we use something like Iceberg.

Reply:
Yes, I thought of the same but that requires spi changes. (isTypeBucketable etc)

Reply:
Also not sure the iceberg connector supports temporary table functionality atm

Reply:
SPI changes could be fine if they're targeted and allow the engine to function properly without baking in underlying presumptions of the storage subsystem.